feat(middleware-code-coverage): allow coverage watermarks configuration via frontend (#145)
SAP/ui5-tooling-extensions: packages/middleware-code-coverage/README.md

        <script src="../../resources/sap/ui/qunit/qunit-junit.js"></script>
        <script src="../../resources/sap/ui/qunit/qunit-coverage-istanbul.js"
          data-sap-ui-cover-only="ui5/sample/"
          data-sap-ui-cover-never="ui5/sample/test/"

>       ></script>

Reviewer: indentation doesn't fit to all other lines

>       ></script>

Reviewer: - indentation doesn't fit to all other lines (same applies to script tag in new section)
- the code above has the same script tag and is not changed, so both are out-of-sync now

}
```

> ### Frontend Configuration

Reviewer 2: ```suggestion
### Front-End Configuration
```

> It is possible to override [`watermarks`](https://github.com/istanbuljs/nyc/blob/ab7c53b2f340b458789a746dff2abd3e2e4790c3/README.md#high-and-low-watermarks) (since UI5 1.119.0) via data attributes in `qunit-coverage-istanbul.js`'s script tag.

Reviewer 2: ```suggestion
You can override [`watermarks`](https://github.com/istanbuljs/nyc/blob/ab7c53b2f340b458789a746dff2abd3e2e4790c3/README.md#high-and-low-watermarks) (since UI5 1.119.0) via data attributes in the script tag for `qunit-coverage-istanbul.js`':
```

```diff html title="unitTests.qunit.html"
...

    <script src="../../resources/sap/ui/qunit/qunit-coverage-istanbul.js"
      data-sap-ui-cover-only="ui5/sample/"
      data-sap-ui-cover-never="ui5/sample/test/"
+     data-sap-ui-cover-watermarks-statements="[90,95]"
+     data-sap-ui-cover-watermarks-functions="[90,95]"
+     data-sap-ui-cover-watermarks-branches="[90,95]"
+     data-sap-ui-cover-watermarks-lines="[90,95]"></script>

...
```

> ## How it works

Reviewer 2: ```suggestion
## How It Works
```

> The middleware adds an HTTP endpoint to the development server. Information about the endpoints can be found in the [API document](./docs/API.md).

Reviewer 2: ```suggestion
The middleware adds an HTTP endpoint to the development server. For more information about the endpoints, see the [API document](./docs/API.md).
```

> The custom middleware intercepts every `.js`-file before it is sent to the client. The file is then instrumented on the fly, including the dynamic creation of a `sourcemap`.

Reviewer 2: ```suggestion
The custom middleware intercepts every `.js`-file before it is sent to the client. The file is then instrumented on the fly, which includes the dynamic creation of a `sourcemap`.
```

The instrumented code and the `sourcemap` are subsequently delivered to the client instead of the original `.js`-file.

## Integration

> The middleware is integrated into OpenUI5 out of the box, but it is not limited just to it. With the [configuration](#configuration) and the [public API](./docs/API.md), developers could set up the middleware to suit their projects' needs.

Reviewer 2: ```suggestion
The middleware is integrated into OpenUI5 out of the box, but you are not limited by this. With the [configuration](#configuration) and the [public API](./docs/API.md), you can set up the middleware to suit your projects' needs.
```

}
```

### Frontend Configuration

> It is possible to override [`watermarks`](https://github.com/istanbuljs/nyc/blob/ab7c53b2f340b458789a746dff2abd3e2e4790c3/README.md#high-and-low-watermarks) (since OpenUI5 1.119.0) via data attributes in `qunit-coverage-istanbul.js`'s script tag.

Reviewer: ```suggestion
It is possible to override [`watermarks`](https://github.com/istanbuljs/nyc/blob/ab7c53b2f340b458789a746dff2abd3e2e4790c3/README.md#high-and-low-watermarks) (since UI5 1.119.0) via data attributes in `qunit-coverage-istanbul.js`'s script tag.
```
We also use `UI5` as a general term already on this page.

```html title="unitTests.qunit.html"
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>Unit tests for OpenUI5 App</title>

    <script id="sap-ui-bootstrap" src="../../resources/sap-ui-core.js"
      data-sap-ui-theme="sap_horizon"
      data-sap-ui-resourceroots='{
        "ui5.sample": "../../"
      }' data-sap-ui-language="EN" data-sap-ui-async="true">
    </script>

    <link rel="stylesheet" type="text/css" href="../../resources/sap/ui/thirdparty/qunit-2.css">

    <script src="../../resources/sap/ui/thirdparty/qunit-2.js"></script>
    <script src="../../resources/sap/ui/qunit/qunit-junit.js"></script>
    <script src="../../resources/sap/ui/qunit/qunit-coverage-istanbul.js"
      data-sap-ui-cover-only="ui5/sample/"
      data-sap-ui-cover-never="ui5/sample/test/"
      data-sap-ui-cover-watermarks-statements="[90,95]"
      data-sap-ui-cover-watermarks-functions="[90,95]"
      data-sap-ui-cover-watermarks-branches="[90,95]"
      data-sap-ui-cover-watermarks-lines="[90,95]"

>   ></script>

Reviewer: Wouldn't it be sufficient and easier to read if we just mention the relevant script tag in here? Everything else is just a copy of the already existing code described above, so nothing that needs to be changed.

Author: Maybe GitHub's Diff markdown will fit here best: https://gist.github.com/salmedina/ad8bea4f46de97ea132f71b0bca73663#file-markdowndiffexample-md

**Note:** Report types could be defined and limited via the middleware's configuration.

> **Note:** Some report settings could be adjusted from the frontend by providing the setting via the request body. Currently, only [`watermarks`](https://github.com/istanbuljs/nyc/blob/ab7c53b2f340b458789a746dff2abd3e2e4790c3/README.md#high-and-low-watermarks) are supported. Frontend defined settings take precedence over default or `ui5.yaml` configured ones.

Reviewer 3: ```suggestion
**Note:** Also it is possible to set report settings from the frontend by providing them via the request body. Currently, only [`watermarks`](https://github.com/istanbuljs/nyc/blob/ab7c53b2f340b458789a746dff2abd3e2e4790c3/README.md#high-and-low-watermarks) are supported. Frontend defined settings take precedence over default or `ui5.yaml` configured ones.
```

**Example:**

```js
fetch("/.ui5/coverage/report", {
  method: "POST",

>   body: JSON.stringify({

Reviewer: Please also update the other example for a request to use the new format.

Author: Thanks!

**Note:** Report types could be defined and limited via the middleware's configuration.

> **Note:** Also it is possible to set report settings from the frontend by providing them via the request body. Currently, only [`watermarks`](https://github.com/istanbuljs/nyc/blob/ab7c53b2f340b458789a746dff2abd3e2e4790c3/README.md#high-and-low-watermarks) are supported (Available since OpenUI5 1.119.0). Frontend defined settings take precedence over default or `ui5.yaml` configured ones.

Reviewer: While this description is not incorrect, it does not really help a user understanding what to do in order to make use of the feature.
The middleware API is not primarily intended to be used directly, so telling that it can be passed as request body is not relevant.

Author: To me, this becomes a little confusing to understand and structure 🤔 

On one side, there's this API documentation where we list the API endpoints with the payload and on the other side, we know that this won't be used standalone, but with Openui5. We acknowledge that we need to somehow also give suggestions on how to use it.
So, how to structure this?

What do you think if I update the HTML samples in the **Usage** section and reference it from here?
